words/source/MovePicker.js

```diff
@@ -0,0 +1,283 @@
+        if (tileArray[i].x == tile.x && tileArray[i].y == tile.y) return true;
+      }
+      return false;
+    }
+    _isMoveValid(move, selectedTiles, tile) {
+      // Initial moves are considered valid.
+      if (
+        !move.coordinates ||
+        move.coordinates.length == 0 ||
+        selectedTiles.length == 0
+      )
+        return true;
+      // Selecting the last selected tile is permitted so as to de-select.
+      let lastSelectedTile = selectedTiles[selectedTiles.length - 1];
+      if (lastSelectedTile.x == tile.x && lastSelectedTile.y == tile.y)
+        return true;
+      // Else the new selection must touch the last selection and can't
+      // already be selected.
+      let touchesLastSelectedTile =
+        (lastSelectedTile.x == tile.x && lastSelectedTile.y == tile.y - 1) ||
+        (lastSelectedTile.x == tile.x && lastSelectedTile.y == tile.y + 1) ||
+        (lastSelectedTile.x == tile.x - 1 && lastSelectedTile.y == tile.y) ||
+        (lastSelectedTile.x == tile.x + 1 && lastSelectedTile.y == tile.y);
+      return (
+        touchesLastSelectedTile &&
+        !this._tileArrayContainsTile(selectedTiles, tile)
+      );
+    }
+    _onTileClicked(e, state) {
+      let tile = new Tile(e.data.value, state.board);
+      let charIndex = e.data.value;
+      let lastSelectedTile =
+        state.selectedTiles.length == 0
+          ? undefined
+          : state.selectedTiles[state.selectedTiles.length - 1];
+      info(
+        `_onTileClicked [tile=${tile.toString}, lastSelectedTile=${
+          lastSelectedTile ? lastSelectedTile.toString : 'undefined'
+        }].`
+      );
+      if (!this._isMoveValid(state.move, state.selectedTiles, tile)) {
+        info(`Ignoring selection of invalid tile [tile=${tile.toString}].`);
+        return;
+      }
+      let newCoordinates = '';
+      if (
+        lastSelectedTile &&
+        lastSelectedTile.x == tile.x &&
+        lastSelectedTile.y == tile.y
+      ) {
+        // User clicked on same tile last clicked, so de-select it.
+        state.selectedTiles.pop();
+        // We could pop the last tuple but it's easier to just rebuild,
+        // and hopefully we'll move away from a string tuple hack soon.
+        for (let i = 0; i < state.selectedTiles.length; i++) {
+          if (i > 0) newCoordinates += ',';
+          let buildTile = state.selectedTiles[i];
+          newCoordinates += `(${buildTile.x},${buildTile.y})`;
+        }
+      } else {
+        // Append the new tile to the existing selection.
+        state.selectedTiles.push(tile);
+        newCoordinates = `(${tile.x},${tile.y})`;
+        if (state.move.coordinates) {
+          newCoordinates = `${state.move.coordinates},${newCoordinates}`;
+        }
+      }
+      state.move.coordinates = newCoordinates;
+      this._setState({
+        board: state.board,
+        move: state.move,
+        selectedTiles: state.selectedTiles
+      });
+    }
+    _onSubmitMove(e, state) {
```

Comment:
I wonder if there's a more descriptive term for 'move'. 'Turn', or 'selected'?

Reply:
I like 'turn', and agree that 'move' isn't a great word. Will look at revising soon.